tools/codegen/api/autograd.py

```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
             if f.func.kind() == SchemaKind.inplace:
                 assert len(info.forward_derivatives) == 1  # Only single output inplace should exist
                 fw_info = info.forward_derivatives[0]
-
-                if re.search(IDENT_REGEX.format("self"), fw_info.formula):
-                    raise RuntimeError(f'The formula for "{f.func.name}" is using the original value of self that is being '
-                                       'modified inplace. This would lead to wrong forward gradients. Please use "result" in '
-                                       'the formula only.')
-
-                # replace "result" from the formula by self
+                formula = fw_info.formula
+
+                def update_for_original(formula: str, postfix: str) -> str:
+                    def repl(m: Match[str]) -> str:
+                        return f'{m.group(1)}original_self{postfix}{m.group(2)}'
+                    return re.sub(IDENT_REGEX.format(f'self{postfix}'), repl, formula)
+
+                if re.search(IDENT_REGEX.format("self_p"), formula):
+                    if is_exact_match:
+                        # For manually defined formulas, don't allow the original value to be used
+                        raise RuntimeError(f'The formula for "{f.func.name}" is using the original value of self '
+                                           'that is being modified inplace. This would lead to wrong forward gradients. '
+                                           'Please use "result" in the formula only.')
+                    else:
+                        # When the original formula is out of place, we save a clone of the primal
+                        # value to be able to access this value if needed
+                        # replace "self_p"/"self_t" from the formula by "original_self_p"/"original_self_t"
+                        formula = update_for_original(formula, "_p")
+                        formula = update_for_original(formula, "_t")
+
+                if re.search(IDENT_REGEX.format("original_self_p"), formula):
+                    required_original_self_value = True
+                else:
+                    required_original_self_value = False
+
+                # replace "result" from the formula by "self_p"
                 def repl(m: Match[str]) -> str:
-                    return f'{m.group(1)}self{m.group(2)}'
-                formula = re.sub(IDENT_REGEX.format("result"), repl, fw_info.formula)
+                    return f'{m.group(1)}self_p{m.group(2)}'
+                formula = re.sub(IDENT_REGEX.format("result"), repl, formula)
+
+                required_primals = fw_info.required_inputs_primal
+                if re.search(IDENT_REGEX.format("self_p"), formula):
```

Comment:
I'm confused: didn't we replace all instances of `self_p` with `original_self_p`?

Reply:
But then we changed all the "result" to "self_p".
Let me add a comment at the beginning explaining the overall logic for this function